consensus/istanbul/celo-ibft-spec.md

```diff
@@ -0,0 +1,488 @@
+`H - height`                    |`Hc - current height`                                      
+`R - round          `           |`Rc - current round`                    
+`V - value`                     |`Rd - desired round`                    
+`T - Message type`              |`Vc - currently proposed value`         
+`RCC - round change certificate`|`PCc - current prepared certificate`    
+`PC - prepared certificate`     |`Sc - current participant state`        
+`S - participant state`         |
+`M or N - message sets`         |
+
+```
+
+// Upon receiving this event the algorithm transitions to the next height the
+// final committed event signifies that the application has accepted the agreed
+// upon value for the current height.
+upon: <FinalCommittedEvent>
+  Hc ← Hc+1
+  Rc ← 0
+  Rd ← 0
+  Sc ← AcceptRequest
+  Vc ← nil
+  schedule onRoundChangeTimeout(Hc, 0) after roundChangeTimeout(0)
+
+// A request event is a request to reach agreement on the provided value, the
+// request event is sent by the application, if this parcicipant is the proposer
+// it will propose that value by sending a preprepared message.
+upon: <RequestEvent, Hc, V> && Sc = AcceptRequest
+  if Rc = 0 && isProposer(Hc, Rd) {
+    bc(<Preprepare, Hc, 0, V, nil>)
+  }
+
+// When receiving a preprepare from a proposer participants will vote for the
+// value (if valid) by sending a prepare message.
+upon: <Preprepare, Hc, Rd, V, RCC> from proposer(Hc, Rd) && Sc = AcceptRequest
+  if (Rd > 0 && validRCC(V, RCC)) || (Rd = 0 && RCC = nil)  {
+    Rc ← Rd
+    Vc ← V
+    Sc ← Preprepared
+    bc(<Prepare, Hc, Rd, Vc>)
+  }
+
+// When a participant sees at least 2f+1 prepare or commit messages for a value
+// it will send a commit message for that value.
+upon: M ← { <T, Hc, Rd, Vc> : T ∈ {Prepare, Commit} } && |M| >= 2f+1 && Sc ∈ {AcceptRequest, Preprepared} 
+  Sc ← Prepared
+  PCc ← <PreparedCertificate, M, Vc>
+  bc(<Commit, Hc, Rd, Vc>)
+
+// When a participant sees at least 2f+1 commit messages for a value, they
+// consider that value committed (agreed) and pass the value to the application,
+// which will in turn issue a final committed event if the value is considered
+// valid by the application.
+upon: M ← { <Commit, Hc, Rd, Vc> } && |M| >= 2f+1 && Sc ∈ {AcceptRequest, Preprepared, Prepared} 
+  Sc ← Committed
+  deliverValue(Vc)
+
+// Upon receipt of a round change if that round change is old, send the participant's round
+// change back to the sender to help them catch up. Othewise if there are at least 2f+1 round
+// change messages sharing the same round then switch to it. Otherwise if there are at least
+// f+1 round change messages switch to the higest round that is less than or equal to the top
+// f+1 rounds.
+upon: m<RoundChange, Hc , R, PC> && (PC = nil || validPC(PC)) 
+  if R < Rd {
+    send(<RoundChange, Hc, Rd, PCc>, sender(m))
+  } else if quorumRound() > Rd {
+    Rd ← quorumRound()
+    Rc ← quorumRound()
+    Sc ← AcceptRequest
```

Comment:
shouldn't we clear the Message sets?

Reply:
Given that they are not defined in this version there is nothing to clear.

Reply:
I don't understand. Are they not part of the state? They seem to be since we add messages to the sets during the protocol execution

Reply:
I'm not sure what message sets you are referring to here?